include/ensmallen_bits/lmcma/lmcma_impl.hpp

```diff
@@ -0,0 +1,298 @@
+/**
+ * @file cmaes_impl.hpp
+ * @author Oleksandr Nikolskyy
+ *
+ * Implementation of the LM CMA algorithm - useful in a derivative-free large-scale
+ * black-box optimization scenario
+ *
+ * For details see "LM-CMA: an Alternative to L-BFGS for Large Scale Black-box Optimization" by Loshchilov
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef ENSMALLEN_LMCMA_LMCMA_IMPL_HPP
+#define ENSMALLEN_LMCMA_LMCMA_IMPL_HPP
+
+#include "lmcma.hpp"
+
+#include <ensmallen_bits/function.hpp>
+
+
+namespace ens {
+    LMCMA::LMCMA(std::size_t N_dim) :
+            T(std::ceil(std::log(N_dim))),
+            c_c(0.5/std::sqrt(N_dim)),
+            c1(1/(10 * std::log(N_dim + 1))),
+            z_bias(0.25),
+            c_sigma(0.3),
+            d_sigma(0.7),
+            m(4 + (size_t)std::floor(3*std::log(N_dim))),
+            lambda(4 + (size_t)std::floor(3*std::log(N_dim))),
+            //lambda(100),
+            w( 1,std::floor( (4 + (size_t)std::floor(3*std::log(N_dim))) / 2 ), arma::fill::zeros),
+            mu(std::floor( (4 + (size_t)std::floor(3 * std::log(N_dim))) / 2 ) )
+    {}
+
+
+    template<typename SeparableFunctionType,
+            typename MatType,
+            typename... CallbackTypes>
+    typename MatType::elem_type LMCMA::optimize(SeparableFunctionType& f,
+                                         MatType& z,
+                                         float sigma,         // TODO: remove from here
+                                         std::size_t n_iter,  // TODO: Remove from here
+                                         CallbackTypes&&... callbacks)
+    {
+
+
+      // Convenience typedefs.
+      typedef typename MatType::elem_type ElemType;
+      typedef typename MatTypeTraits<MatType>::BaseMatType BaseMatType;
+
+      // Make sure that we have the methods that we need.  Long name...
+      traits::CheckArbitrarySeparableFunctionTypeAPI<
+              SeparableFunctionType, BaseMatType>();
+      RequireDenseFloatingPointType<BaseMatType>();
+
+      BaseMatType& iterate = (BaseMatType&) z;
```

Comment:
Can we instead use ```iterateIn``` instead of z? Unless it has a symbolic meaning in the actual paper. See [cne_impl.hpp](https://github.com/mlpack/ensmallen/blob/1b6c1d93f09adc2313c67827355e6f9c9f958d3c/include/ensmallen_bits/cne/cne_impl.hpp#L89)